php-checks/src/main/java/org/sonar/php/checks/regex/AnchorPrecedenceCheck.java

```diff
@@ -0,0 +1,117 @@
+ */
+package org.sonar.php.checks.regex;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonarsource.analyzer.commons.regex.RegexParseResult;
+import org.sonarsource.analyzer.commons.regex.ast.BoundaryTree;
+import org.sonarsource.analyzer.commons.regex.ast.DisjunctionTree;
+import org.sonarsource.analyzer.commons.regex.ast.NonCapturingGroupTree;
+import org.sonarsource.analyzer.commons.regex.ast.RegexBaseVisitor;
+import org.sonarsource.analyzer.commons.regex.ast.RegexTree;
+import org.sonarsource.analyzer.commons.regex.ast.SequenceTree;
+
+@Rule(key = "S5850")
+public class AnchorPrecedenceCheck extends AbstractRegexCheck {
+
+  public static final String MESSAGE = "Group parts of the regex together to make the intended operator precedence explicit.";
+
+  @Override
+  public void checkRegex(RegexParseResult regexParseResult, FunctionCallTree regexFunctionCall) {
+    new Visitor().visit(regexParseResult);
+  }
+
+  private enum Position {
+    BEGINNING, END
+  }
+
+  private class Visitor extends RegexBaseVisitor {
+
+    @Override
+    public void visitDisjunction(DisjunctionTree tree) {
+      List<RegexTree> alternatives = tree.getAlternatives();
+      if ((anchoredAt(alternatives, Position.BEGINNING) || anchoredAt(alternatives, Position.END))
+        && notAnchoredElseWhere(alternatives)) {
+        newIssue(tree, MESSAGE);
+      }
+      super.visitDisjunction(tree);
+    }
+
+    private boolean anchoredAt(List<RegexTree> alternatives, Position position) {
+      int itemIndex = position == Position.BEGINNING ? 0 : (alternatives.size() - 1);
+      RegexTree firstOrLast = alternatives.get(itemIndex);
+      return isAnchored(firstOrLast, position);
+    }
+
+    private boolean notAnchoredElseWhere(List<RegexTree> alternatives) {
+      if (isAnchored(alternatives.get(0), Position.END)
+        || isAnchored(alternatives.get(alternatives.size() - 1), Position.BEGINNING)) {
+        return false;
+      }
+      for (RegexTree alternative : alternatives.subList(1, alternatives.size() - 1)) {
+        if (isAnchored(alternative, Position.BEGINNING) || isAnchored(alternative, Position.END)) {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private boolean isAnchored(RegexTree tree, Position position) {
+      if (!tree.is(RegexTree.Kind.SEQUENCE)) {
+        return false;
+      }
+      SequenceTree sequence = (SequenceTree) tree;
+      List<RegexTree> items = sequence.getItems().stream()
+        .filter(item -> !isFlagSetter(item))
+        .collect(Collectors.toList());
+      if (items.isEmpty()) {
+        return false;
+      }
+      int index = position == Position.BEGINNING ? 0 : (items.size() - 1);
+      RegexTree firstOrLast = items.get(index);
+      return firstOrLast.is(RegexTree.Kind.BOUNDARY) && isAnchor((BoundaryTree) firstOrLast);
```

Comment:
We don't check whether the boundary type makes sense for the position. We end up raising issues on stupid code like:
```
    preg_match('/$a|b|c/', $input);
    preg_match('/a|b|c^/', $input);
```
As it's stupid code and as the issue message isn't really wrong, it can be acceptable.

Reply:
Indeed, those regexes don't make much sense. I think, however, it is still valuable to inform about the precedence here. For example:
```php
<?php
preg_match('/foo|bar|car^/', 'foo'); // returns true. Even if the ^ was probably meant to be a $, maybe the user did only test with `foo` and `bar`
```
On those regexes we would also raise S5996 `ImpossibleBoundariesCheck`. If the user in `a|b|c^` just replaces the `^` with $ to fix S5996, he'll get `S5850` raised anyway in the next scan.